mesonbuild/mparser.py

```diff
@@ -20,6 +20,30 @@
 
 # This is the regex for the supported escape sequences of a regular string
 # literal, like 'abc\x00'
+
+# Backslash escaping happens in several levels/places in Python and
+# works differently in each place. Survival skills:
+#
+# - By design, a backslash in a _raw_ string NEVER gets combined with
+#   the next character. Always use raw strings for regular
+#   expressions. Note even when never combined, it still has a bit of
+#   "escape power" as seen in: r'a\'b'
+#
+# - A regular expression ALWAYS combines a backslash with the next
+#   character. So finding a literal backslash requires doubling
+#   it. Starting with 3.6, sequences with an unknown letter (e.g: r'\z')
+#   fail to compile. Other unknown sequences discard the backslash as
+#   no-op, e.g: r'\=' matches exactly like r'='.
+#
+# - A backslash is treated the SAME inside vs outside a regex set [ ].
+#   However '\+' makes a difference outside [ ]\+, whereas the same
+#   backslash is discarded as a no-op inside [\+] because + is not
+#   special inside.
+#
+# - Regular, not-raw strings combine backslashes with the next character
+#   SOMETIMES and sometimes not; check the relevant documentation. In
```

Comment:
Not, sometimes, always. The cases where it doesn't cause warnings and will be errors in the future.

Reply:
Read the next paragraph:

>    Changed in version 3.6: Unrecognized escape sequences produce a `DeprecationWarning`. In some future version of Python they will be a `SyntaxError`.

You should not encourage unescaped backslashes.

I don't know why you mention Meson's syntax when this comment is about Python.

Reply:
> You should not encourage unescaped backslashes.

Good thing I'm not :-)

> I don't know why you mention Meson's syntax when this comment is about Python.

Various reasons but I'd rather not "review the review". In fact I didn't expect this particular PR to be reviewed at all, see its description and draft status. Thanks anyway!

Reply:
BTW this does not show any `DeprecationWarning`, any idea why?
```
printf '%s\n' "s='a\zb'; print(s)" > unknownescape.py
python3.7 unknownescape.py 
python3.7 -m compileall .
```